components/support/sql/src/conn_ext.rs

```diff
@@ -213,3 +217,72 @@ impl<'conn> ConnExt for UncheckedTransaction<'conn> {
         &*self
     }
 }
+
+/// This transaction is suitable for when a transaction is being used only
+/// for performance when updating a larger number of rows (ie, when the
+/// transaction is not necessary for data integrity). You can specify a
+/// duration for the maximum amount of time the transaction should be held.
+/// You should regularly call .maybe_commit() as part of your processing, and
+/// if the current transaction has been open for greater than the specified
+/// duration the transaction will be committed and another one started.
+/// You should always call .commit() at the end. Note that if you call
+/// .rollback() at the end, only the current transaction will be rolled back.
+pub struct TransactionForPerf<'conn> {
+    tx: UncheckedTransaction<'conn>,
+    behavior: TransactionBehavior,
+    commit_after: Duration,
+}
+
+impl<'conn> TransactionForPerf<'conn> {
+    /// Begin a new transaction which may be split into multiple transactions
+    /// for performance reasons. Cannot be nested, but this is not
+    /// enforced; use a rusqlite `savepoint` for nested transactions.
+    pub fn new(
+        conn: &'conn Connection,
+        behavior: TransactionBehavior,
+        commit_after: Duration,
+    ) -> SqlResult<Self> {
+        Ok(Self {
+            tx: UncheckedTransaction::new(conn, behavior)?,
+            behavior,
+            commit_after,
+        })
+    }
+
+    /// Checks to see if we have held a transaction for longer than the
+    /// requested time, and if so, commits the current transaction and opens
+    /// another.
+    pub fn maybe_commit(&mut self) -> SqlResult<()> {
+        if self.tx.started_at.elapsed() >= self.commit_after {
```

Comment:
In the future it might be worth putting the body of this (e.g. what we do if the time is elapsed) in an #[inline(never)] function, and then inline the rest of this function (but this is only likely to matter if it starts getting called inside tight loops or something).